crates/project_panel/src/project_panel.rs

```diff
@@ -449,30 +455,53 @@ impl ProjectPanel {
     }
 
     fn update_diagnostics(&mut self, cx: &mut ViewContext<Self>) {
-        self.diagnostic_error_paths = self
-            .project
+        let show_warnings =
+            ProjectPanelSettings::get_global(cx).show_diagnostics == ShowDiagnostics::All;
+        self.diagnostics = Default::default();
+        self.project
             .read(cx)
             .diagnostic_summaries(false, cx)
             .filter_map(|(path, _, diagnostic_summary)| {
                 if diagnostic_summary.error_count > 0 {
-                    Some(path)
+                    Some((path, DiagnosticSeverity::ERROR))
+                } else if show_warnings && diagnostic_summary.warning_count > 0 {
+                    Some((path, DiagnosticSeverity::WARNING))
                 } else {
                     None
                 }
             })
-            .flat_map(|project_path| {
-                let mut path_buf = PathBuf::new();
-                project_path
-                    .path
-                    .components()
-                    .map(|component| {
-                        path_buf.push(component);
-                        path_buf.clone()
-                    })
-                    .chain(std::iter::once(PathBuf::from("")))
-                    .collect::<Vec<_>>()
+            .for_each(|(project_path, diagnostic_severity)| {
+                let mut path_buffer = PathBuf::new();
+                self.update_highest_diagnostic_severity(
+                    &project_path,
+                    path_buffer.clone(),
+                    diagnostic_severity,
+                );
```

Comment:
This here is necessary to also include the root directory of the worktree. We need the empty buffer in the HashMap as well.